bmips: add new target

Noltari opened the pull request:
bmips features full device tree support, as opposed to bcm63xx.

~~It relies on mtd for reading CFE mac address instead of memory mapped nvram, which makes it easier to add support for devices with NAND and other bootloaders (u-boot, redboot).~~

~~Only BCM6328 is supported for now, as a first approach. Support for other SoCs can be added later.~~

Comtrend AR-5387un bootlog:
```
[    0.000000] Linux version 5.10.16 (noltari@skynet) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 8.4.0 r15882-aeb74bbf0e) 8.4.0, GNU ld (GNU Binutils) 2.34) #0 SMP Mon Feb 22 11:27:39 2021
[    0.000000] CPU0 revision is: 0002a075 (Broadcom BMIPS4350)
[    0.000000] MIPS: machine is Comtrend AR-5387un
[    0.000000] earlycon: bcm63xx_uart0 at MMIO 0x10000100 (options '115200n8')
[    0.000000] printk: bootconsole [bcm63xx_uart0] enabled
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 16 bytes.
[    0.000000] Primary data cache 32kB, 2-way, VIPT, cache aliases, linesize 16 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] percpu: Embedded 15 pages/cpu s29200 r8192 d24048 u61440
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: console=ttyS0,115200n8 earlycon
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 50012K/65536K available (5781K kernel code, 624K rwdata, 1172K rodata, 7008K init, 207K bss, 15524K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=128, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu: 	RCU restricting CPUs from NR_CPUS=2 to nr_cpu_ids=1.
[    0.000000] 	Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=1
[    0.000000] NR_IRQS: 256
[    0.000000] irq_bcm6345_l1: registered BCM6345 L1 intc (IRQs: 64)
[    0.000000] irq_bcm6345_l1:   CPU0 at MMIO 0x(ptrval) (irq = 2)
[    0.000000] random: get_random_bytes called from start_kernel+0x3d8/0x604 with crng_init=0
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 11945377789 ns
[    0.000022] sched_clock: 32 bits at 160MHz, resolution 6ns, wraps every 13421772796ns
[    0.008186] Calibrating delay loop... 319.74 BogoMIPS (lpj=639488)
[    0.050571] pid_max: default: 32768 minimum: 301
[    0.055777] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.063377] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.078710] rcu: Hierarchical SRCU implementation.
[    0.084678] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.093212] smp: Bringing up secondary CPUs ...
[    0.097949] smp: Brought up 1 node, 1 CPU
[    0.109246] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.119378] futex hash table entries: 256 (order: 3, 32768 bytes, linear)
[    0.126687] pinctrl core: initialized pinctrl subsystem
[    0.133655] NET: Registered protocol family 16
[    0.343418] 10000100.serial: ttyS0 at MMIO 0x10000100 (irq = 28, base_baud = 1562500) is a bcm63xx_uart
[    0.353224] printk: console [ttyS0] enabled
[    0.353224] printk: console [ttyS0] enabled
[    0.361855] printk: bootconsole [bcm63xx_uart0] disabled
[    0.361855] printk: bootconsole [bcm63xx_uart0] disabled
[    0.374785] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.405082] clocksource: Switched to clocksource MIPS
[    0.414000] NET: Registered protocol family 2
[    0.420882] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.429726] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.437743] TCP bind hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.445157] TCP: Hash tables configured (established 1024 bind 1024)
[    0.452338] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.459294] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.467304] NET: Registered protocol family 1
[    0.878051] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.895769] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.901853] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.919141] bcm6328-pinctrl 10000080.pin-controller: registered at mmio (ptrval)
[    0.930612] bcm63xx-power-controller 10001848.power-controller: registered 10 power domains
[    0.940571] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    1.020990] spi-nor spi1.0: mx25l12805d (16384 Kbytes)
[    1.067012] 3 fixed-partitions partitions found on MTD device spi1.0
[    1.073694] Creating 3 MTD partitions on "spi1.0":
[    1.078679] 0x000000000000-0x000000010000 : "cfe"
[    1.089031] 0x000000010000-0x000000ff0000 : "firmware"
[    1.105177] mtdsplit_bcm63xx: CFE image tag found at 0x0 with version 6, board type 96328A-1441N1
[    1.114513] random: fast init done
[    1.118774] 2 bcm63xx-fw partitions found on MTD device firmware
[    1.125067] Creating 2 MTD partitions on "firmware":
[    1.130220] 0x000000000100-0x00000023b45c : "kernel"
[    1.135341] mtd: partition "kernel" doesn't start on an erase/write block boundary -- force read-only
[    1.150421] 0x00000023b45c-0x000000fe0000 : "rootfs"
[    1.155655] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
[    1.172939] mtd: device 3 (rootfs) set to be root filesystem
[    1.182221] 1 squashfs-split partitions found on MTD device rootfs
[    1.188808] 0x000000420000-0x000000fe0000 : "rootfs_data"
[    1.200149] 0x000000ff0000-0x000001000000 : "nvram"
[    1.215930] libphy: Fixed MDIO Bus: probed
[    1.221990] bcm7038-wdt 1000005c.watchdog: Registered BCM7038 Watchdog
[    1.234079] NET: Registered protocol family 10
[    1.270543] Segment Routing with IPv6
[    1.274741] NET: Registered protocol family 17
[    1.279501] 8021q: 802.1Q VLAN Support v1.8
[    1.396952] Freeing unused kernel memory: 7008K
[    1.401650] This architecture does not have kernel memory protection.
[    1.408300] Run /init as init process
[    2.315340] init: Console is alive
[    2.319512] init: - watchdog -
[    2.362769] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.414903] usbcore: registered new interface driver usbfs
[    2.420874] usbcore: registered new interface driver hub
[    2.426622] usbcore: registered new device driver usb
[    2.458250] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    2.468533] ehci-fsl: Freescale EHCI Host controller driver
[    2.481161] ehci-platform: EHCI generic platform driver
[    2.487159] bcm63xx_usbh_phy_xlate: 0
[    2.491035] bcm63xx_usbh_phy_xlate: args_count=0
[    2.495793] bcm63xx_usbh_phy_xlate: args=2170075884
[    3.543609] ehci-platform 10002500.usb: EHCI Host Controller
[    3.549540] ehci-platform 10002500.usb: new USB bus registered, assigned bus number 1
[    3.558125] ehci-platform 10002500.usb: irq 42, io mem 0x10002500
[    3.576724] ehci-platform 10002500.usb: USB 2.0 started, EHCI 1.00, overcurrent ignored
[    3.587677] hub 1-0:1.0: USB hub found
[    3.594344] hub 1-0:1.0: 1 port detected
[    3.622133] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    3.632015] ohci-platform: OHCI generic platform driver
[    3.638044] bcm63xx_usbh_phy_xlate: 0
[    3.641893] bcm63xx_usbh_phy_xlate: args_count=0
[    3.646649] bcm63xx_usbh_phy_xlate: args=2170075884
[    3.652085] ohci-platform 10002600.usb: Generic Platform OHCI controller
[    3.659110] ohci-platform 10002600.usb: new USB bus registered, assigned bus number 2
[    3.667578] ohci-platform 10002600.usb: irq 41, io mem 0x10002600
[    3.739371] hub 2-0:1.0: USB hub found
[    3.746075] hub 2-0:1.0: 1 port detected
[    3.755225] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.773769] init: - preinit -
[    4.405942] random: jshn: uninitialized urandom read (4 bytes read)
[    4.503520] random: jshn: uninitialized urandom read (4 bytes read)
[    4.556561] random: jshn: uninitialized urandom read (4 bytes read)
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    8.981498] procd: - early -
[    8.984836] procd: - watchdog -
[    9.630427] procd: - watchdog -
[    9.634642] procd: - ubus -
[    9.658940] urandom_read: 2 callbacks suppressed
[    9.658958] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.689277] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.697047] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.710688] procd: - init -
Please press Enter to activate this console.
[   11.222822] kmodloader: loading kernel modules from /etc/modules.d/*
[   11.419437] xt_time: kernel timezone is -0000
[   11.516847] PPP generic driver version 2.4.2
[   11.534785] NET: Registered protocol family 24
[   11.613127] kmodloader: done loading kernel modules from /etc/modules.d/*
[   11.750492] urngd: v1.0.2 started.
[   12.065909] random: crng init done
[   12.069482] random: 2 urandom warning(s) missed due to ratelimiting
```

adrianschmutzler commented on target/linux/bmips/Makefile:
```diff
@@ -0,0 +1,19 @@
+BOARDNAME:=Broadcom BMIPS
+CPU_TYPE:=mips32
+FEATURES:=squashfs usb pci
+KERNEL_PATCHVER:=5.4
```
Consider adding an empty line before so KERNEL_PATCHVER is separate from the block.

adrianschmutzler commented on target/linux/bmips/base-files/etc/uci-defaults/09_fix_crc:
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
uci-defaults files are sourced, so the should not contain a shebang.

However, it is convenient to add "exit 0" at the end of the file, because otherwise there deletion will depend on the status of the latest command.

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+/dts-v1/;
```
Typically, we request SPDX license identifiers for new device support PRs.

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+};
+
+&ehci {
+	status = "ok";
```
"okay"? (Is ok still working on 5.4?)

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+&hsspi {
+	status = "okay";
+
+	m25p80@0 {
```
flash@0 ?

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+};
+
+&leds {
+	status = "ok";
```
"okay" ...

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+		     &pinctrl_ephy0_act_led &pinctrl_ephy1_act_led
+		     &pinctrl_ephy2_act_led &pinctrl_ephy3_act_led>;
+
+	inet_red: led@1 {
```
Personally, I'd use the full "internet" instead of "inet" for DT label and label property.

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+
+	inet_red: led@1 {
+		reg = <1>;
+		label = "AR-5387un:red:inet";
```
Don't know whether the label is constrained by something, but if you start from scratch it would be convenient to use the model part of the device name here: "ar-5387un:red:internet"
This has the advantage that it can be derived from device name if necessary.

adrianschmutzler commented on target/linux/bmips/dts/bcm6328-comtrend-ar-5387un.dts:
```diff
@@ -0,0 +1,212 @@
+};
+
+&ohci {
+	status = "ok";
```
"okay" here and below ...

adrianschmutzler commented on target/linux/bmips/dts/bcm6328.dtsi:
```diff
@@ -0,0 +1,450 @@
+#include <dt-bindings/clock/bcm6328-clock.h>
```
License?

adrianschmutzler commented on target/linux/bmips/dts/bcm6328.dtsi:
```diff
@@ -0,0 +1,450 @@
+			status = "disabled";
+		};
+
+		enet: ethernet@10e00000 {
```
Any reason for the abbreviated names? I'd just have used "ethernet" like on ramips ...

adrianschmutzler commented on target/linux/bmips/image/Makefile:
```diff
@@ -0,0 +1,116 @@
+	LOADADDR=$(KERNEL_LOADADDR) \
+	KERNEL_ENTRY=$(KERNEL_ENTRY) \
+	RAMSIZE=$(RAMSIZE) \
+	LZMA_TEXT_START=$(LZMA_TEXT_START) \
```
This final line-wrap "\" intended?

Noltari replied:
This was inherited from bcm63xx :$

adrianschmutzler commented on target/linux/bmips/image/Makefile:
```diff
@@ -0,0 +1,116 @@
+  SOC = bcm$$(CHIP_ID)
+  DEVICE_DTS = $$(SOC)-$(subst _,-,$(1))
+endef
+DEVICE_VARS += CHIP_ID SOC
```
SOC is already added to DEVICE_VARS in the common script.

adrianschmutzler commented on target/linux/bmips/image/bcm63xx.mk:
```diff
@@ -0,0 +1,33 @@
+  CFE_EXTRAS := --block-size $$(BLOCK_SIZE) --image-offset $$(if $$(IMAGE_OFFSET),$$(IMAGE_OFFSET),$$(BLOCK_SIZE))
+endef
+DEVICE_VARS += BLOCK_SIZE FLASH_MB IMAGE_OFFSET
+DEVICE_VARS += CFE_BOARD_ID CFE_EXTRAS
```
Consider putting those to the start of the file like introduced in:
https://github.com/openwrt/openwrt/commit/5b392c7119f23aef10ffc0def9a70e8335f9f3f8

adrianschmutzler commented on target/linux/bmips/image/bcm63xx.mk:
```diff
@@ -0,0 +1,33 @@
+  CFE_BOARD_ID :=
+  CFE_EXTRAS := --block-size $$(BLOCK_SIZE) --image-offset $$(if $$(IMAGE_OFFSET),$$(IMAGE_OFFSET),$$(BLOCK_SIZE))
+endef
+DEVICE_VARS += BLOCK_SIZE FLASH_MB IMAGE_OFFSET
```
Most targets use "BLOCKSIZE" which is defined globally in https://github.com/openwrt/openwrt/blob/master/include/image.mk (and added to DEVICE_VARS there).

I wonder whether we can switch to the common name here? And if not, maybe then using a more specific name BLOCKSIZE_CFE or similar would be more appropriate?

adrianschmutzler commented on target/linux/bmips/image/bcm63xx.mk:
```diff
@@ -0,0 +1,33 @@
+  IMAGE/sysupgrade.bin := cfe-bin
+  BLOCK_SIZE := 0x10000
+  IMAGE_OFFSET :=
+  FLASH_MB := 4
```
I do not think it's a good idea to provide a default for that value. From my experience at other targets, variables like this are frequently overlooked, and then just people set nothing and inherit the 4 MiB for a device with different size (don't know whether that's technically possible here).

So, I'd put the `FLASH_MB :=` initializer here and then require to set the actual value  for _every_ device.

dengqf6 commented on target/linux/bmips/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,19 @@
+
+case "$(board_name)" in
+comtrend,ar-5387un)
+	ucidef_add_switch "switch0" \
```
Since this is a new target, I suggest use upstream B53 DSA driver instead of swconfig.

Noltari replied:
@dengqf6 unfortunately, dsa b53 and bcm63xx-enetsw drivers needs some modifications to get it working, and I would need help from @KanjiMonster / @ffainelli for that...

Noltari replied:
Marking this as resolved since I've removed downstream ethernet drivers.

hauke commented:
Would it be possible to replace the BCM63xx target with this one and what would be needed?

Noltari commented:
@hauke the enetsw driver needs some love, but that's out of my knowledge.
The older enet driver (6338, 6345, 6348, 6358) would also need to be converted to device tree.
I already have the BCM6318 PCIe driver working, but I couldn't get the older PCI driver working with bmips yet (6338, 6345, 6348, 6358, 6368).
However, it's probably reasonable to drop support for 6338, 6345 and 6348 devices, since most of them have small flashes and RAM.

ffainelli commented:
> 
> 
> @hauke the enetsw driver needs some love, but that's out of my knowledge.

For enetsw, it would be nice if this could use the DSA b53 driver.

> The older enet driver (6338, 6345, 6348, 6358) would also need to be converted to device tree.
> I already have the BCM6318 PCIe driver working, but I couldn't get the older PCI driver working with bmips yet (6338, 6345, 6348, 6358, 6368).
> However, it's probably reasonable to drop support for 6338, 6345 and 6348 devices, since most of them have small flashes and RAM.

Yes that sounds reasonable to me.

Noltari commented:
> > @hauke the enetsw driver needs some love, but that's out of my knowledge.
> 
> For enetsw, it would be nice if this could use the DSA b53 driver.

Yes, it would be really niced, but I've tried and I couldn't make it work.
Maybe you can help with that?

Noltari commented:
PR updated with:
- Generic/NAND subtargets.
- BCM6318 support.
- BCM6328 support.
- BCM6362 support.
- NAND support.
- Fixed bug in BCM6362 pinctrl.
- BCM6328 LED driver improvement.
- LZMA loader improvements.

adrianschmutzler commented on package/kernel/bcm63xx-cferam/Makefile:
```diff
@@ -0,0 +1,42 @@
+include $(INCLUDE_DIR)/kernel.mk
+
+PKG_NAME:=bcm63xx-cferam
+PKG_VERSION:=2020-05-10
```
Is this really a version or might it be more appropriate to use PKG_SOURCE_DATE?

Noltari replied:
You're right, I'll change it

adrianschmutzler commented on package/kernel/bcm63xx-cferam/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_SOURCE_VERSION:=bc48a2657ec5a4a3fec88bb6d3bedca5de6c062f
+PKG_MIRROR_HASH:=d584cbfb15acacb8e64cbc93e55663e43aa769955580f364d57a1135402143b7
+
+PKG_BUILD_DIR:=$(KERNEL_BUILD_DIR)/$(PKG_NAME)
```
If you don't use PKG_VERSION, I think this can be dropped.

Noltari replied:
Ok

adrianschmutzler commented on target/linux/bmips/base-files/etc/board.d/02_network:
```diff
@@ -0,0 +1,24 @@
+	ucidef_add_switch "switch0" \
+		"0:lan:1" "1:lan:2" "2:lan:3" "3:lan:4" "8t@eth0"
+	;;
+comtrend,ar-5315u)
```
Should be moved to the top due to sorting ...

adrianschmutzler commented on target/linux/bmips/image/bcm63xx_generic.mk:
```diff
@@ -0,0 +1,51 @@
+  BLOCKSIZE := 0x10000
+  IMAGE_OFFSET :=
+  FLASH_MB :=
+  CHIP_ID :=
```
That's already defined in Device/Default

adrianschmutzler commented on target/linux/bmips/image/bcm63xx_nand.mk:
```diff
@@ -0,0 +1,71 @@
+  IMAGE/cfe.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | cfe-jffs2-cferam | append-ubi | cfe-wfi-tag
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  KERNEL_SIZE := 5120k
+  CHIP_ID :=
```
Already defined in Device/Default.

adrianschmutzler commented:
Sorry for only posting nitpick comments BTW.

If it starts to annoy you, just tell me and I will stop :-)

Noltari commented:
> Sorry for only posting nitpick comments BTW.
> 
> If it starts to annoy you, just tell me and I will stop :-)

Your suggestions are always welcome :)

Noltari commented:
PR updated with BCM6368 support.
Note: SMP support disabled, since enetsw crashes when SMP is enabled. I need to investigate why this happens only in this specific SoC...

Noltari committed:
bcm63xx-cferam: select for bmips target

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
kernel: mtdsplit: add support for BCM63XX CFE firmware

Broadcom CFE bootloader relies on a tag for identifying the current firmware,
such as version, image start address, kernel address and size, rootfs size,
board id, signatures, etc.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
mtd: fixtrx: support CFE imagetag on bmips target

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
PR updated:
- Solved SMP issues with BCM6368 enetsw driver.
- Switched to @KanjiMonster interrupt controller, since it seems to be more stable than upstream controlller (at least with BCM6358).

Noltari commented:
PR updated with upstreamed patches:
- SPI and HSSPI.
- USBH PHY.
- Gated clock improvements (and BCM6318 support).
The number of downstream patches has been considerably reduced. However, the following patches should be upstream:
- IRQs.
- watchdog.
- pinctrl.
- ethernet.
- PCI.

Noltari committed:
bmips: add new target

This target has full device tree support, thus reducing the number of
patches needed for bcm63xx, in which there's a patch for every board.

The intention is to start with a minimal amount of downstream patches and
start upstreaming all of them.

Current status:
 - Enabling EHCI/OHCI on BCM6358 causes a kernel panic.
 - BCM63268 lacks Timer Clocks/Reset support.
 - No PCI/PCIe drivers.
 - No ethernet drivers.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>
Acked-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

Noltari commented:
Rebased on latest OpenWrt master and removed a considerable amount of downstream patches.
bmips target is now source only and the intention is to have as few downstream patches as possible, so I will start upstreaming the current downstream patches and then we can start working on the currently non-supported features (watchdog, pci, ethernet...).

@adschm can you take a look at this?
I'm a big fan of your NITs reviews :)

Noltari commented:
Added watchdog support.

Noltari commented:
Fixed small bug in mtdsplit (pr_debug instead of pr_err).

Noltari commented:
Fixed USB support on BCM6318, BCM6328, BCM6362 and BCM63268 (except usb_ref clock).

adschm reviewed (commented):
I scanned through part of the DTS files, so don't except me to have checked everything there. Most of the DTS comments apply to other files as well, and I have only put them once.

I have not looked at all at lzma-loader and kernel patches so far, and I'm not sure whether I will have time for that soon.

adschm commented:
On the latest virtual meeting, it was mentioned that probably all new files should get an SPDX license identifier in the first place (though there was no decision of any kind that this would be formally "mandatory"). Consider adding appropriate ones for the scripts (board.d/lib) and the Makefiles.  If unsure, the project license GPL-2.0-only or GPL-2.0-or-later should be chosen. Obviously, it would not make much sense to apply this to kernel-config or patches.

adschm commented:
Are the patches 200-300 upstreamable (i.e. is this just not been brought into proper shape yet or is it not planned to be upstreamed)?

adschm commented:
If you plan to upstream the DTS files eventually as well, consider to add the proper color/function properties already, so you don't have to go through all files again at a later point just for that reason. However, since we do not support that properly in OpenWrt yet, you would still be required to have the label property there as well (though, it's technically redundant).

Note that this "consider" really means "consider", I'm not entirely sure which way is the right one to go here (not doing it now would have the advantage that the feature might have matured when you implement it eventually). But if you are touching all these files now anyway, you should at least briefly think about what you prefer in this context ...

Noltari commented on target/linux/bmips/base-files/etc/board.d/01_leds:
```diff
@@ -0,0 +1,19 @@
+board_config_update
+
+led_ethernet="$(get_dt_led ethernet)"
+[ -n "$led_ethernet" ] && ucidef_set_led_netdev "lan" "lan" "$led_ethernet" "eth0"
```
I will remove this since ethernet isn't supported (yet).

Noltari commented on target/linux/bmips/files/include/dt-bindings/interrupt-controller/bcm6318-interrupt-controller.h:
```diff
@@ -0,0 +1,84 @@
+#define BCM6318_IRQ_USB_BULK_TX_DMA	8
+#define BCM6318_IRQ_USB_ISO_RX_DMA	9
+#define BCM6318_IRQ_USB_ISO_TX_DMA	10
+#define BCM6318_IRQ_DG			11
```
It's not wrong, it's just Github's viewer.

Noltari commented on target/linux/bmips/image/Makefile:
```diff
@@ -0,0 +1,266 @@
+DEVICE_VARS += CHIP_ID DEVICE_LOADADDR
+
+USB1_PACKAGES := kmod-usb-ohci kmod-ledtrig-usbdev
+USB2_PACKAGES := $(USB1_PACKAGES) kmod-usb2
```
I think this is personal, and I prefer to use those defines since all the boards are going to use the same packages USBs are exposed.

Noltari commented:
> Are the patches 200-300 upstreamable (i.e. is this just not been brought into proper shape yet or is it not planned to be upstreamed)?

Yes, these patches should be easily upstreamable.
1xx and 4xx require some work to be upstreamed.

Noltari commented:
> If you plan to upstream the DTS files eventually as well, consider to add the proper color/function properties already, so you don't have to go through all files again at a later point just for that reason. However, since we do not support that properly in OpenWrt yet, you would still be required to have the label property there as well (though, it's technically redundant).
> 
> Note that this "consider" really means "consider", I'm not entirely sure which way is the right one to go here (not doing it now would have the advantage that the feature might have matured when you implement it eventually). But if you are touching all these files now anyway, you should at least briefly think about what you prefer in this context ...



> If you plan to upstream the DTS files eventually as well, consider to add the proper color/function properties already, so you don't have to go through all files again at a later point just for that reason. However, since we do not support that properly in OpenWrt yet, you would still be required to have the label property there as well (though, it's technically redundant).
> 
> Note that this "consider" really means "consider", I'm not entirely sure which way is the right one to go here (not doing it now would have the advantage that the feature might have matured when you implement it eventually). But if you are touching all these files now anyway, you should at least briefly think about what you prefer in this context ...

I think I'll wait, since I have no intention in upstreaming the DTS files.

Noltari commented:
Many thanks for your review @adschm!
All you comments should have been addressed with my latest push.

Noltari commented on target/linux/bmips/dts/bcm63168-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,197 @@
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
```
Actually we don't need them, thanks for pointing that out.

Noltari commented on target/linux/bmips/dts/bcm63168-comtrend-vr-3032u.dts:
```diff
@@ -0,0 +1,197 @@
+
+		reset {
+			label = "reset";
+			gpios = <&pinctrl 33 1>;
```
I've also replaced 0 for GPIO_ACTIVE_HIGH.

Noltari commented on target/linux/bmips/image/Makefile:
```diff
@@ -0,0 +1,266 @@
+
+ifeq ($(SUBTARGET),nand)
+include bcm63xx_nand.mk
+endif
```
Yeah, since it's possible to add support for other SoCs (bcm33xx and bcm7xxx), I think it's better to use what you suggested.

adschm commented on target/linux/bmips/dts/bcm6362-netgear-dgnd3700-v2.dts:
```diff
@@ -0,0 +1,225 @@
+			};
+
+			partition@1d00000 {
+				label = "xxx";
```
That's for porn?

Noltari replied:
Porn from OEM firmware :P
See https://openwrt.org/toh/netgear/dgnd3700v2#oem_bootlog
```
Creating 14 MTD partitions on "brcmnand.0":
0x000000080000-0x000001b40000 : "rootfs"
0x000000000000-0x000000004000 : "nvram"
0x000001b40000-0x000001b80000 : "language_DEU"
0x000001dc0000-0x000001ec0000 : "scnvram"
0x000001ec0000-0x000001f00000 : "factory"
0x000000080000-0x000001cc0000 : "upgrade"
0x000001c80000-0x000001cc0000 : "flag"
0x000001cc0000-0x000001d00000 : "pcbasn"
0x000001d00000-0x000001d80000 : "xxx"
0x000001b80000-0x000001bc0000 : "language_PTB"
0x000001bc0000-0x000001c00000 : "language_RUS"
0x000001c00000-0x000001c40000 : "language_CHS"
0x000001c40000-0x000001c80000 : "language_ENU"
0x000001d80000-0x000001dc0000 : "language_dev"
```

adschm replied:
Well, then let's keep there default-porn read-only and forget about this :-)

adschm commented on target/linux/bmips/image/Makefile:
```diff
@@ -0,0 +1,262 @@
+  DEVICE_DTS = $$(SOC)-$(subst _,-,$(1))
+  DEVICE_LOADADDR :=
+endef
+DEVICE_VARS += CHIP_ID DEVICE_LOADADDR
```
Consider moving DEVICE_VARS to the top, like in the *.mk files or here:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/Makefile#L6

adschm commented on target/linux/bmips/image/bcm63xx_generic.mk:
```diff
@@ -0,0 +1,78 @@
+
+#
+# BCM63xx Profiles
+#
```
These comments can probably be dropped.

adschm commented on target/linux/bmips/image/bcm63xx_nand.mk:
```diff
@@ -0,0 +1,85 @@
+
+#
+# BCM63xx NAND Profiles
+#
```
See above, I personally would remove this comment. But again, taste.

adschm reviewed (commented):
Another set of nitpicks ...

Noltari commented on target/linux/bmips/dts/bcm6358-huawei-hg556a-b.dts:
```diff
@@ -0,0 +1,178 @@
+};
+
+/*
+FIXME! Enabling EHCI/OHCI causes a kernel panic
```
Yes, I just wanted to make sure that nobody enabled that until the bug is fixed.

Noltari committed:
image: add ModelNameLimit16

This script returns the model name limited to 16 characters.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
bcm63xx: limit name metadata to model name

Since there are only 16 characters available, on most cases the vendor name
will fit in the metadata, but the model name won't fit.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

rikka0w0 commented:
> Could it be installed on a Sagemcom F@ST 3686? Because I see that it uses the BCM3384ZU chipset.

I successfully ran OpenWRT on a Sagemcom F@st 3864OP router (marked as unsupported). The router has a BCM63168D0 dual-core CPU, an internal switch, and an external BCM53124S switch. [My port](https://github.com/rikka0w0/openwrt-fast3864op/tree/3864_latest) is based on the latest master commit (bc43ad88ed18722c0621fd6dfef0ff68268f4e14), as of 2022 07 21. I saw the BMIPS branch fully supports DTS and the external MDIO bus, so the port is based on BMIPS target instead of BCM63XX. At this moment, only the latest master can be used, as the other branches don't have one important patch:
[target/linux/bmips/patches-5.15/501-net-dsa-b53-mmap-fix-enabled_ports.patch](https://github.com/openwrt/openwrt/blob/9b4169df2d26ddcb8cea41de47c159d421389130/target/linux/bmips/patches-5.15/501-net-dsa-b53-mmap-fix-enabled_ports.patch)

The image boots if I only enable the internal switch. With this setup, the WAN port is functional, but none of the LAN ports works. With the help of the `mdio` tool, I can read the phy status of both switches. However, once I enable the external switch, the system hangs during boot:

```
[    2.086792] bcm6368-mdio-mux 107000b0.mdio: Broadcom BCM6368 MDIO mux bus
[    2.095453] bcm53xx 0.1:1e: found switch: BCM53125, rev 4
[    2.101076] bcm53xx 0.1:1e: failed to register switch: -517
[    2.107601] b53-switch 10700000.switch: found switch: BCM63xx, rev 0
[    2.118387] bcm6368-enetsw 1000d800.ethernet: mtd mac d8:d7:75:13:03:72
[    2.248248] bcm7038-wdt 1000009c.watchdog: Registered BCM7038 Watchdog
[    2.261817] NET: Registered PF_INET6 protocol family
[    2.298322] Segment Routing with IPv6
[    2.302295] In-situ OAM (IOAM) with IPv6
[    2.306643] NET: Registered PF_PACKET protocol family
[    2.312635] 8021q: 802.1Q VLAN Support v1.8
[    2.349723] bcm53xx 0.1:1e: found switch: BCM53125, rev 4
[    2.355488] bcm53xx 0.1:1e: failed to register switch: -517
[    2.361827] b53-switch 10700000.switch: found switch: BCM63xx, rev 0
^ Stucked at this line
```

If I disable the internal switch, then the external switch can be successfully recognized. Both switches use the b53 driver, but they cannot co-exist at this moment. [My DTS](https://github.com/rikka0w0/openwrt-fast3864op/blob/7aeec3d1c1fd413c2d526c6f736a8ab25d72cce1/target/linux/bmips/dts/bcm63168-sagem-fast-3864op.dts)
